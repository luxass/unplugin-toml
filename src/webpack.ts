/**
 * This entry file is for webpack plugin.
 *
 * @module
 */

import type unplugin from "./";
import { createWebpackPlugin } from "unplugin";
import { unpluginFactory } from "./";

/**
 * Webpack plugin
 *
 * @example
 * ```ts
 * // webpack.config.js
 * module.exports = {
 *  plugins: [require("unplugin-toml/webpack")()],
 * }
 * ```
 */
export default createWebpackPlugin(unpluginFactory) as typeof unplugin.webpack;
