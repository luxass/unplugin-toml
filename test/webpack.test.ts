import type { Configuration } from "webpack";
import { join } from "node:path";
import { dedent } from "@luxass/utils";
import { describe, expect, it } from "vitest";
import { testdir } from "vitest-testdirs";
import { webpack as createWebpack } from "webpack";
import TOMLPlugin from "../src/webpack";

async function webpack(config: Configuration, testdirPath: string): Promise<null> {
  return new Promise((resolve, reject) => {
    const compiler = createWebpack({
      optimization: {
        minimize: true,
        usedExports: false,
      },
      experiments: {
        outputModule: true,
      },
      output: {
        library: {
          type: "module",
        },
        path: join(testdirPath, "dist"),
        filename: "bundle.js",
      },
      mode: "production",
      ...config,
    });

    if (compiler == null) {
      throw new Error("webpack compiler is not defined");
    }

    compiler.run((err, stats) => {
      if (err) {
        reject(err);
      }

      if (!stats) {
        reject(new Error("webpack stats not available"));
        return;
      }

      resolve(null);
    });
  });
}

describe("webpack", () => {
  it("expect toml import to be a json object", async () => {
    const testdirPath = await testdir.from(join(import.meta.dirname, "fixtures/basic"));

    expect(testdirPath).toBeDefined();

    await webpack({
      entry: join(testdirPath, "basic.js"),
      plugins: [
        TOMLPlugin(),
      ],
    }, testdirPath);

    const config = await import(join(testdirPath, "dist/bundle.js")).then((m) => m.config);
    expect(config).toBeDefined();

    expect(config).toEqual({
      pluginDir: "./plugins",
      web: { enabled: true },
      logging: { type: "stdout", level: "info" },
    });
  });

  it("expect toml import to be a string", async () => {
    const testdirPath = await testdir.from(join(import.meta.dirname, "fixtures/basic"));

    expect(testdirPath).toBeDefined();

    await webpack({
      entry: join(testdirPath, "basic-raw.js"),
      plugins: [
        TOMLPlugin(),
      ],
    }, testdirPath);

    const config = await import(join(testdirPath, "dist/bundle.js")).then((m) => m.config);
    expect(config).toBeDefined();

    expect(config).toMatch(dedent`
      pluginDir = "./plugins"

      [web]
      enabled = true

      [logging]
      type = "stdout"
      level = "info"
    `);
  });

  it("handle transforms", async () => {
    const testdirPath = await testdir.from(join(import.meta.dirname, "fixtures/transform"));

    expect(testdirPath).toBeDefined();

    await webpack({
      entry: join(testdirPath, "transform.js"),
      plugins: [
        TOMLPlugin({
          transform(data) {
            if (data != null && typeof data === "object" && "this" in data) {
              return {
                this: "transformed",
              };
            }
          },
        }),
      ],
    }, testdirPath);

    const config = await import(join(testdirPath, "dist/bundle.js")).then((m) => m.config);
    expect(config).toBeDefined();

    expect(config).toEqual({
      this: "transformed",
    });
  });
});
