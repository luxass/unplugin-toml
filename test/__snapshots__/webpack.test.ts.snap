// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handle transforms 1`] = `"(()=>{"use strict";console.log({this:"transformed"})})();"`;

exports[`handles toml > expect toml import to be a json object 1`] = `"(()=>{"use strict";console.log({pluginDir:"./plugins",web:{enabled:!0},logging:{type:"stdout",level:"info"}})})();"`;

exports[`handles toml > expect toml import to be a string 1`] = `"(()=>{"use strict";console.log('pluginDir = "./plugins"\\n\\n[web]\\nenabled = true\\n\\n[logging]\\ntype = "stdout"\\nlevel = "info"\\n')})();"`;
